server/s3_put_multiobject_copy_action.cc

```diff
@@ -0,0 +1,962 @@
+    old_probable_del_rec.reset(new S3ProbableDeleteRecord(
+        old_oid_rec_key, {0ULL, 0ULL},
+        object_multipart_metadata->get_object_name(), old_object_oid,
+        old_layout_id, object_multipart_metadata->get_pvid_str(),
+        bucket_metadata->get_multipart_index_layout().oid,
+        bucket_metadata->get_objects_version_list_index_layout().oid, "",
+        false /* force_delete */, true,
+        part_metadata->get_part_index_layout().oid, 0,
+        strtoul(part_metadata->get_part_number().c_str(), NULL, 0),
+        bucket_metadata->get_extended_metadata_index_layout().oid));
+
+    probable_oid_list[old_oid_rec_key] = old_probable_del_rec->to_json();
+  }
+  // prepending a char depending on the size of the object (size based bucketing
+  // of object)
+  S3CommonUtilities::size_based_bucketing_of_objects(
+      new_oid_str, request->get_content_length());
+
+  s3_log(S3_LOG_DEBUG, request_id,
+         "Adding new_probable_del_rec with key [%s]\n", new_oid_str.c_str());
+  new_probable_del_rec.reset(new S3ProbableDeleteRecord(
+      new_oid_str, old_object_oid, object_multipart_metadata->get_object_name(),
+      new_object_oid, layout_id, object_multipart_metadata->get_pvid_str(),
+      bucket_metadata->get_multipart_index_layout().oid,
+      bucket_metadata->get_objects_version_list_index_layout().oid, "",
+      false /* force_delete */, true,
+      part_metadata->get_part_index_layout().oid, 1,
+      strtoul(part_metadata->get_part_number().c_str(), NULL, 0),
+      bucket_metadata->get_extended_metadata_index_layout().oid));
+  // store new oid, key = newoid
+  probable_oid_list[new_oid_str] = new_probable_del_rec->to_json();
+
+  if (!motr_kv_writer) {
+    motr_kv_writer =
+        motr_kv_writer_factory->create_motr_kvs_writer(request, s3_motr_api);
+  }
+
+  motr_kv_writer->put_keyval(
+      global_probable_dead_object_list_index_layout, probable_oid_list,
+      std::bind(&S3PutMultipartCopyAction::next, this),
+      std::bind(&S3PutMultipartCopyAction::
+                     add_object_oid_to_probable_dead_oid_list_failed,
+                this));
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutMultipartCopyAction::add_object_oid_to_probable_dead_oid_list_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  s3_copy_part_action_state = S3CopyPartActionState::probableEntryRecordFailed;
+  if (motr_kv_writer->get_state() == S3MotrKVSWriterOpState::failed_to_launch) {
+    set_s3_error("ServiceUnavailable");
+  } else {
+    set_s3_error("InternalError");
+  }
+  // Clean up will be done after response.
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutMultipartCopyAction::send_response_to_s3_client() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  if (S3Option::get_instance()->is_getoid_enabled()) {
+
+    request->set_out_header_value("x-stx-oid",
```

Comment:
> x-stx-oid
x-stx-layout-id

These headers looks seagate specific, whats the intent for these @knrajnambiar76 @DPG17 ? Do we need these for UploadPartCopy also?

Reply:
It wont be needed I feel, multipart completion can send the same

Reply:
These are needed for debugging purpose and for our STs. They are custom headers in the response, only used by our System tests (STs) and debugging. The x-stx-oid would be the newly created part object

Reply:
Thanks for clarification. Keeping these headers.